webui-src/app/files/files_util.js

```diff
@@ -91,212 +91,140 @@ function calcRemainingTime(bytes, rate) {
-      break;
-
     case 'pause':
-      actionHeader = '/rsFiles/FileControl';
       jsonParams.flags = RS_FILE_CTRL_PAUSE;
       break;
 
     case 'resume':
-      actionHeader = '/rsFiles/FileControl';
       jsonParams.flags = RS_FILE_CTRL_START;
       break;
 
     case 'force_check':
-      actionHeader = '/rsFiles/FileControl';
       jsonParams.flags = RS_FILE_CTRL_FORCE_CHECK;
       break;
 
     default:
       console.error('Unknown action in Downloads.control()');
       return;
   }
-  const res = await rs.rsJsonApiRequest(actionHeader, jsonParams, () => {});
-  return res.body.retval;
-}
-
-function actionButton(file, action) {
-  switch (action) {
-    case 'resume':
-      return m(
-        'button',
-        {
-          title: 'resume',
-
-          onclick() {
-            fileAction(file.hash, 'resume');
-          },
-        },
-        m('i.fas.fa-play')
-      );
-
-    case 'pause':
-      return m(
-        'button',
-        {
-          title: 'pause',
-
-          onclick() {
-            fileAction(file.hash, 'pause');
-          },
-        },
-        m('i.fas.fa-pause')
-      );
-
-    case 'cancel':
-      return m(
-        'button.red',
-        {
-          title: 'cancel',
-
-          onclick() {
-            widget.popupMessage(
-              m('Cancelpop', [
-                m('p', 'Are you sure you want to cancel download?'),
-                m(
-                  'button',
-                  {
-                    onclick: async () => {
-                      const res = await fileAction(file.hash, 'cancel');
-                      if (res) {
-                        widget.popupMessage(m('p', 'Download Cancelled Successfully'));
-                      } else {
-                        widget.popupMessage(m('p', 'Download Cancel Failed'));
-                      }
-
-                      m.redraw();
-                    },
-                  },
-                  'Cancel'
-                ),
-              ])
-            );
-            // fileAction(file.hash, 'cancel');
-          },
-        },
-        m('i.fas.fa-times')
-      );
-  }
+  rsJsonApiRequest('/rsFiles/FileControl', jsonParams);
 }
 
 const ProgressBar = () => {
   return {
     view: (v) =>
-      m('.progressbar', [
-        m('span.progressbar-status', {
-          style: {
-            width: v.attrs.rate + '%',
-          },
-        }),
-        m('span.progressbar-percent', v.attrs.rate.toPrecision(3) + '%'),
+      m('.progress-bar-chunks', [
+        v.attrs.chunksInfo.chunks.map((item) => m(`span.chunk[data-chunkVal=${item}]`)),
+        m('span.progress-bar-chunks__percent', v.attrs.rate.toPrecision(3) + '%'),
       ]),
   };
 };
 
-const chunkStrats = {
-  0: 'Streaming', // CHUNK_STRATEGY_STREAMING
-  1: 'Random', // CHUNK_STRATEGY_RANDOM
-  2: 'Progressive', // CHUNK_STRATEGY_PROGRESSIVE
-};
-// rstypes.h :: 366
-
 const File = () => {
   let chunkStrat;
+  let rsJsonApiRequest = () => {};
```

Comment:
Why is this needed?

Reply:
At that time I thought to initialise the function first and then assign it with then assign it with the passed value.

Reply:
The rswebui function is undefined here because there is a circular import happening. The rswebui file is importing a files_util function which is causing files_util to initialize before rswebui does.

Reply:
> The rswebui function is undefined here because there is a circular import happening. The rswebui file is importing a files_util function which is causing files_util to initialize before rswebui does.

What to do here?

Reply:
Multiple options, you could maybe create a new files_proxy.js to move all the proxyObj related stuff there, so imports aren't circular. Or you could generalize the create proxy util and move it to rswebui itself

Reply:
> Multiple options, you could maybe create a new files_proxy.js to move all the proxyObj related stuff there, so imports aren't circular. Or you could generalize the create proxy util and move it to rswebui itself

I have tried both, but the circular import is happening in either ways. I am pushing the code, please guide me.

Reply:
> Multiple options, you could maybe create a new files_proxy.js to move all the proxyObj related stuff there, so imports aren't circular. Or you could generalize the create proxy util and move it to rswebui itself

I have got an idea, I was thinking of exposing eventQueue object in rswebui.js, but seems like events property of the object is already exported, So I am going to try to use it in the files_util.js itself.

Reply:
Alright, It works now. No circular imports. Please review @rottencandy